File: src/main/java/site/gachontable/gachontablebe/domain/user/usecase/UserLoginImpl.java
package site.gachontable.gachontablebe.domain.user.usecase;

import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import site.gachontable.gachontablebe.domain.shared.Role;
import site.gachontable.gachontablebe.domain.shared.exception.PasswordNotMatchException;
import site.gachontable.gachontablebe.domain.user.domain.User;
import site.gachontable.gachontablebe.domain.user.domain.repository.UserRepository;
import site.gachontable.gachontablebe.domain.user.exception.UserNotFoundException;
import site.gachontable.gachontablebe.global.jwt.JwtProvider;
import site.gachontable.gachontablebe.global.jwt.dto.JwtResponse;
import site.gachontable.gachontablebe.global.jwt.exception.ExpiredTokenException;
import site.gachontable.gachontablebe.global.jwt.exception.InvalidTokenException;

@Service
@RequiredArgsConstructor
public class UserLoginImpl implements UserLogin{

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtProvider jwtProvider;

    @Override
    public JwtResponse execute(String username, String password) {
        User user = userRepository.findByUsername(username).orElseThrow(UserNotFoundException::new);
        validatePassword(password, user);

        String accessToken = jwtProvider.generateAccessToken(user.getUserId(), user.getUsername(), Role.ROLE_USER);
        String refreshToken = generateRefreshToken(user);

        return new JwtResponse(accessToken, refreshToken);
    }

    private void validatePassword(String password, User user) {
        if (!passwordEncoder.matches(password, user.getPassword())) {
            throw new PasswordNotMatchException();
        }
    }

    private String generateRefreshToken(User user) {
        String refreshToken = user.getRefreshToken();
        if (refreshToken == null || !isValidToken(refreshToken)) {
            refreshToken = jwtProvider.generateRefreshToken(user.getUserId(), user.getUsername(), Role.ROLE_USER);
            updateRefreshToken(user, refreshToken);
        }
        return refreshToken;
    }

    private boolean isValidToken(String token) {
        try {
            jwtProvider.validateToken(token);
            return true;
        } catch (InvalidTokenException | ExpiredTokenException e) {
            return false;
        }
    }

    private void updateRefreshToken(User user, String refreshToken) {
        user.updateRefreshToken(refreshToken);
        userRepository.save(user);
    }
}

File: src/main/java/site/gachontable/domain/member/service/UserLoginImpl.kt
package site.gachontable.domain.member.service

import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import site.gachontable.domain.member.domain.User
import site.gachontable.domain.member.exception.UserNotFoundException
import site.gachontable.domain.member.port.`in`.UserLogin
import site.gachontable.domain.member.port.out.UserRepository
import site.gachontable.infra.security.jwt.JwtProvider
import site.gachontable.infra.security.jwt.dto.JwtResponse
import site.gachontable.presentation.shared.Role
import site.gachontable.presentation.shared.exception.PasswordNotMatchException

@Service
class UserLoginImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtProvider: JwtProvider,
) : UserLogin {
    override fun execute(id: String, password: String): JwtResponse {
        val user = userRepository.findByUsername(id)
            ?: throw UserNotFoundException()

        validatePassword(password, user)

        val accessToken = jwtProvider.generateAccessToken(user.userId, user.userTel, Role.ROLE_USER)
        val refreshToken = generateRefreshToken(user)

        return JwtResponse(accessToken, refreshToken)
    }

    private fun validatePassword(password: String, user: User) {
        if (!passwordEncoder.matches(password, user.password)) {
            throw PasswordNotMatchException()
        }
    }

    private fun generateRefreshToken(user: User): String {
        var refreshToken = user.refreshToken
        if (refreshToken == null || !jwtProvider.isValidToken(refreshToken)) {
            refreshToken = jwtProvider.generateRefreshToken(user.userId, user.username, Role.ROLE_USER)
            updateRefreshToken(user, refreshToken)
        }
        return refreshToken
    }

    private fun updateRefreshToken(user: User, refreshToken: String) {
        user.updateRefreshToken(refreshToken)
        userRepository.save(user)
    }
}
